HIP/hip-0000-atomic-association-contracts.md

```diff
@@ -0,0 +1,104 @@
+---
+hip: 0000
+title: Waived association requirement for atomic transfers for smart contracts
+author: Matthew DeLorenzo <mdelore.ufl@gmail.com>, Vae Vecturne <vea.vecturne@gmail.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2023-04-06
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/696
+---
+
+## Abstract
+
+Smart contracts are required to be associated with all tokens they custody. This makes sense when the contract has a long lived balance > 0; Hedera state is expanded and must be paid for in ongoing maintenance and storage costs to the network. However, there are many use cases that call for smart contracts atomically taking custody of a token, often to transfer to another entity within the same contract call. In this case, association is still required but Hedera state is not expanded.
+
+We propose that association not be required for Hedera entities that only custody a token during an atomic transaction.
+
+## Motivation
+
+Currently, a smart contract attempting to custody an unassociated token will be immediately rejected by the network. This presents a problem in some decentralized finance (DeFi) applications that use smart contracts to atomically custody tokens. Decentraliztion may be achieved by allowing contracts to remain token agnostic, but also prevent spamming attacks by allowing any user to associate tokens to a DeFi smart contract. 
+
+## Rationale
+
+Periphery contracts for decentralized exchanges (DEXes), such as Uniswap V2 and V3, are designed to atomically custody tokens. For DEXes on Hedera, gas-inefficient workarounds are used to associate and dissociate tokens to periphery smart contracts, paid by end-users.
+
+## Specification
+
+Hedera Smart Contract Service optimistically allows custody of any token for smart contracts, and checks at the end of a contract call that tokens are associated to entities whose balance is changing from zero to nonzero. 
```

Comment:
Would this feature be opt-in or opt-out?
In either case if applicable what procedure would a contract creator take?